File: src/test/java/example/sort/service/TestHeapsortService.java
package example.sort.service;
import otterop.lang.Array;
import otterop.lang.String;
import otterop.test.Test;
import otterop.test.TestBase;

public class TestHeapsortService extends TestBase {

    @Test
    public void normalSort() {
        String item;
        Array<String> items = Array.newArray(3, (String) null);
        items.set(0, String.wrap("3"));
        items.set(1, String.wrap("2"));
        items.set(2, String.wrap("1"));
        SortService sortService = new HeapsortService();
        Array<String> itemsSorted = sortService.sort(items);
        this.assertTrue(itemsSorted.size() == 3, String.wrap("itemsSorted should have size 3"));
        item = itemsSorted.get(0);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0,
                String.wrap("expected 1 as first element"));
        item = itemsSorted.get(1);
        this.assertTrue(item.compareTo(String.wrap("2")) == 0,
                String.wrap("expected 2 as second element"));
        item = itemsSorted.get(2);
        this.assertTrue(item.compareTo(String.wrap("3")) == 0,
                String.wrap("expected 3 as third element"));
    }

    @Test
    public void singleItem() {
        String item;
        Array<String> items = Array.newArray(1, (String) null);
        items.set(0, String.wrap("1"));
        SortService sortService = new HeapsortService();
        Array<String> itemsSorted = sortService.sort(items);
        this.assertTrue(itemsSorted.size() == 1, String.wrap("itemsSorted should have size 1"));
        item = itemsSorted.get(0);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0,
                String.wrap("expected 1 as first element"));
    }

    @Test
    public void sameItem() {
        String item;
        Array<String> items = Array.newArray(3, (String) null);
        items.set(0, String.wrap("1"));
        items.set(1, String.wrap("1"));
        items.set(2, String.wrap("1"));

        SortService sortService = new HeapsortService();
        Array<String> itemsSorted = sortService.sort(items);
        this.assertTrue(itemsSorted.size() == 3, String.wrap("itemsSorted should have size 3"));
        item = itemsSorted.get(0);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0,
                String.wrap("expected 1 as first element"));
        item = itemsSorted.get(1);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0,
                String.wrap("expected 1 as second element"));
        item = itemsSorted.get(2);
        this.assertTrue(item.compareTo(String.wrap("1")) == 0,
                String.wrap("expected 1 as third element"));
    }
}

File: dotnet/tests/Example/Sort/Service/TestHeapsortService.cs
namespace Example.Sort.Service
{
    public class TestHeapsortService : Otterop.Test.TestBase
    {
        [Otterop.Test.Test]
        public void NormalSort()
        {
            Otterop.Lang.String item;
            Otterop.Lang.Array<Otterop.Lang.String> items = Otterop.Lang.Array.NewArray(3, (Otterop.Lang.String?) null);
            items.Set(0, Otterop.Lang.String.Wrap("3"));
            items.Set(1, Otterop.Lang.String.Wrap("2"));
            items.Set(2, Otterop.Lang.String.Wrap("1"));
            SortService sortService = new HeapsortService();
            Otterop.Lang.Array<Otterop.Lang.String> itemsSorted = sortService.Sort(items);
            AssertTrue(itemsSorted.Size() == 3, Otterop.Lang.String.Wrap("itemsSorted should have size 3"));
            item = itemsSorted.Get(0);
            AssertTrue(item.CompareTo(Otterop.Lang.String.Wrap("1")) == 0, Otterop.Lang.String.Wrap("expected 1 as first element"));
            item = itemsSorted.Get(1);
            AssertTrue(item.CompareTo(Otterop.Lang.String.Wrap("2")) == 0, Otterop.Lang.String.Wrap("expected 2 as second element"));
            item = itemsSorted.Get(2);
            AssertTrue(item.CompareTo(Otterop.Lang.String.Wrap("3")) == 0, Otterop.Lang.String.Wrap("expected 3 as third element"));
        }

        [Otterop.Test.Test]
        public void SingleItem()
        {
            Otterop.Lang.String item;
            Otterop.Lang.Array<Otterop.Lang.String> items = Otterop.Lang.Array.NewArray(1, (Otterop.Lang.String?) null);
            items.Set(0, Otterop.Lang.String.Wrap("1"));
            SortService sortService = new HeapsortService();
            Otterop.Lang.Array<Otterop.Lang.String> itemsSorted = sortService.Sort(items);
            AssertTrue(itemsSorted.Size() == 1, Otterop.Lang.String.Wrap("itemsSorted should have size 1"));
            item = itemsSorted.Get(0);
            AssertTrue(item.CompareTo(Otterop.Lang.String.Wrap("1")) == 0, Otterop.Lang.String.Wrap("expected 1 as first element"));
        }

        [Otterop.Test.Test]
        public void SameItem()
        {
            Otterop.Lang.String item;
            Otterop.Lang.Array<Otterop.Lang.String> items = Otterop.Lang.Array.NewArray(3, (Otterop.Lang.String?) null);
            items.Set(0, Otterop.Lang.String.Wrap("1"));
            items.Set(1, Otterop.Lang.String.Wrap("1"));
            items.Set(2, Otterop.Lang.String.Wrap("1"));
            SortService sortService = new HeapsortService();
            Otterop.Lang.Array<Otterop.Lang.String> itemsSorted = sortService.Sort(items);
            AssertTrue(itemsSorted.Size() == 3, Otterop.Lang.String.Wrap("itemsSorted should have size 3"));
            item = itemsSorted.Get(0);
            AssertTrue(item.CompareTo(Otterop.Lang.String.Wrap("1")) == 0, Otterop.Lang.String.Wrap("expected 1 as first element"));
            item = itemsSorted.Get(1);
            AssertTrue(item.CompareTo(Otterop.Lang.String.Wrap("1")) == 0, Otterop.Lang.String.Wrap("expected 1 as second element"));
            item = itemsSorted.Get(2);
            AssertTrue(item.CompareTo(Otterop.Lang.String.Wrap("1")) == 0, Otterop.Lang.String.Wrap("expected 1 as third element"));
        }

    }

}
